# Analyze CIA factbook Data Using SQL
# Introduction:
In this project we work with [CIA World Factbook](https://github.com/fivethirtyeight/data/blob/master/college-majors/recent-grads.csv) data. It is a compendium of statistics about all of the countries in Earth. The Factbook contains demographic information like population, population_growth, area etc.

Connecting the jupyter notebook to database :

In [1]:
%%capture
%load_ext sql
%sql sqlite:///factbook.db

# Overview of the Data
We'll begin by getting a sense of what the data looks like.

In [2]:
%%sql
SELECT *
  FROM facts
 LIMIT 5;

 * sqlite:///factbook.db
Done.


[(1, 'af', 'Afghanistan', 652230, 652230, 0, 32564342, 2.32, 38.57, 13.89, 1.51),
 (2, 'al', 'Albania', 28748, 27398, 1350, 3029278, 0.3, 12.92, 6.58, 3.3),
 (3, 'ag', 'Algeria', 2381741, 2381741, 0, 39542166, 1.84, 23.67, 4.31, 0.92),
 (4, 'an', 'Andorra', 468, 468, 0, 85580, 0.12, 8.13, 6.96, 0.0),
 (5, 'ao', 'Angola', 1246700, 1246700, 0, 19625353, 2.78, 38.78, 11.49, 0.46)]

Here are the descriptions for some of the columns:

* name - The name of the country.
* area - The total land and sea area of the country.
* population - The country's population.
* population_growth- The country's population growth as a percentage.
* birth_rate - The country's birth rate, or the number of births a year per 1,000 people.
* death_rate - The country's death rate, or the number of death a year per 1,000 people.
* area- The country's total area (both land and water).
* area_land - The country's land area in square kilometers.
* area_water - The country's waterarea in square kilometers.

Let's start by calculating some summary statistics and see what they tell us.

# Summary Statistics 
Displaying following values from the database in a single query.

* min_population - minimum population
* max_population - maximum population
* min_pop_growth - minimum population growth
* max_pop_growth - maximum population growth

In [3]:
%%sql
SELECT MIN(population) AS min_pop,
       MAX(population) AS max_pop,
       MIN(population_growth) AS min_pop_growth,
       MAX(population_growth) max_pop_growth 
  FROM facts;

 * sqlite:///factbook.db
Done.


[(0, 7256490011, 0.0, 4.02)]

A few things stick out from the summary statistics in the last screen:

* There's a country with a population of 0
* There's a country with a population of 7256490011 (or more than 7.2 billion people)

Let's use subqueries to zoom in on just these countries without using the specific values.

# Exploring Outliers 
Retrieving Country with minimum population :

In [5]:
%%sql
SELECT *
  FROM facts
 WHERE population == (SELECT MIN(population)
                        FROM facts
                     );

 * sqlite:///factbook.db
Done.


[(250, 'ay', 'Antarctica', None, 280000, None, 0, None, None, None, None)]

It seems like the table contains a row for Antarctica, which explains the population of 0. This seems to match the CIA Factbook [page for Antarctica](https://www.cia.gov/library/publications/the-world-factbook/geos/ay.html):

<img src = "https://s3.amazonaws.com/dq-content/257/fb_antarctica.png">

Retrieving Country with maximum population:

In [6]:
%%sql
SELECT *
  FROM facts
 WHERE population == (SELECT MAX(population)
                        FROM facts
                     );

 * sqlite:///factbook.db
Done.


[(261, 'xx', 'World', None, None, None, 7256490011, 1.08, 18.6, 7.8, None)]

We also see that the table contains a row for the whole world, which explains the maximum population of over 7.2 billion we found earlier.

Now that we know this, we should recalculate the summary statistics we calculated earlier, while excluding the row for the whole world.

# Summary statistics Revisited

In [7]:
%%sql
SELECT MIN(population) AS min_pop,
       MAX(population) AS max_pop,
       MIN(population_growth) AS min_pop_growth,
       MAX(population_growth) AS max_pop_growth 
  FROM facts
 WHERE name <> 'World';

 * sqlite:///factbook.db
Done.


[(0, 1367485388, 0.0, 4.02)]

There's a country whose population closes in on 1.4 billion!

# Exploring Average Population and Area
Let's explore density. Density depends on the population and the country's area. Let's look at the average values for these two columns.

We should take care of discarding the row for the whole planet.

In [8]:
%%sql
SELECT AVG(population) AS avg_population, AVG(area) AS avg_area
  FROM facts
 WHERE name <> 'World';

 * sqlite:///factbook.db
Done.


[(32242666.56846473, 555093.546184739)]

We see that the average population is around 32 million and the average area is 555 thousand square kilometers.

# Finding Densely Populated Countries
To finish, we'll build on the query above to find countries that are densely populated. We'll identify countries that have:

* Above average values for population.
* Below average values for area.

In [9]:
%%sql
SELECT *
  FROM facts
 WHERE population > (SELECT AVG(population)
                       FROM facts
                    )
   AND area < (SELECT AVG(area)
                 FROM facts
);

 * sqlite:///factbook.db
Done.


[(14, 'bg', 'Bangladesh', 148460, 130170, 18290, 168957745, 1.6, 21.14, 5.61, 0.46),
 (65, 'gm', 'Germany', 357022, 348672, 8350, 80854408, 0.17, 8.47, 11.42, 1.24),
 (85, 'ja', 'Japan', 377915, 364485, 13430, 126919659, 0.16, 7.93, 9.51, 0.0),
 (138, 'rp', 'Philippines', 300000, 298170, 1830, 100998376, 1.61, 24.27, 6.11, 2.09),
 (173, 'th', 'Thailand', 513120, 510890, 2230, 67976405, 0.34, 11.19, 7.8, 0.0),
 (185, 'uk', 'United Kingdom', 243610, 241930, 1680, 64088222, 0.54, 12.17, 9.35, 2.54),
 (192, 'vm', 'Vietnam', 331210, 310070, 21140, 94348835, 0.97, 15.96, 5.93, 0.3)]

Some of these countries are generally known to be densely populated, so we have confidence in our results!

# Countries having the highest ratios of water to land

In [10]:
%%sql
SELECT *, CAST(area_water as float)/CAST(area_land as float) water_to_land_ratio FROM facts
WHERE water_to_land_ratio = 
(SELECT 
MAX(CAST(area_water as float)/CAST(area_land as float))
FROM facts);

 * sqlite:///factbook.db
Done.


[(228, 'io', 'British Indian Ocean Territory', 54400, 60, 54340, None, None, None, None, None, 905.6666666666666)]

# Countries have more water than land

In [11]:
%%sql
SELECT * FROM facts 
WHERE 
area_water > area_land

 * sqlite:///factbook.db
Done.


[(228, 'io', 'British Indian Ocean Territory', 54400, 60, 54340, None, None, None, None, None),
 (247, 'vq', 'Virgin Islands', 1910, 346, 1564, 103574, 0.59, 10.31, 8.54, 7.67)]

# countries will add the most people to their population next year
Displaying countries having more than 3% population growth by next year.

In [12]:
%%sql
SELECT * FROM facts
WHERE
population_growth > 3;

 * sqlite:///factbook.db
Done.


[(27, 'uv', 'Burkina Faso', 274200, 273800, 400, 18931686, 3.03, 42.03, 11.72, 0.0),
 (29, 'by', 'Burundi', 27830, 25680, 2150, 10742276, 3.28, 42.01, 9.27, 0.0),
 (106, 'mi', 'Malawi', 118484, 94080, 24404, 17964697, 3.32, 41.56, 8.41, 0.0),
 (128, 'ng', 'Niger', None, 1266700, 300, 18045729, 3.25, 45.45, 12.42, 0.56),
 (141, 'qa', 'Qatar', 11586, 11586, 0, 2194817, 3.07, 9.84, 1.53, 22.39),
 (162, 'od', 'South Sudan', 644329, None, None, 12042910, 4.02, 36.91, 8.18, 11.47),
 (182, 'ug', 'Uganda', 241038, 197100, 43938, 37101745, 3.24, 43.79, 10.69, 0.74)]

# countries have a higher death rate than birth rate

In [15]:
%%sql
SELECT name, death_rate, birth_rate FROM facts 
WHERE
death_rate>birth_rate

 * sqlite:///factbook.db
Done.


[('Austria', 9.42, 9.41),
 ('Belarus', 13.36, 10.7),
 ('Bosnia and Herzegovina', 9.75, 8.87),
 ('Bulgaria', 14.44, 8.92),
 ('Croatia', 12.18, 9.45),
 ('Czech Republic', 10.34, 9.63),
 ('Estonia', 12.4, 10.51),
 ('Germany', 11.42, 8.47),
 ('Greece', 11.09, 8.66),
 ('Hungary', 12.73, 9.16),
 ('Italy', 10.19, 8.74),
 ('Japan', 9.51, 7.93),
 ('Latvia', 14.31, 10.0),
 ('Lithuania', 14.27, 10.1),
 ('Moldova', 12.59, 12.0),
 ('Monaco', 9.24, 6.65),
 ('Poland', 10.19, 9.74),
 ('Portugal', 11.02, 9.27),
 ('Romania', 11.9, 9.14),
 ('Russia', 13.69, 11.6),
 ('Serbia', 13.66, 9.08),
 ('Slovenia', 11.37, 8.42),
 ('Ukraine', 14.46, 10.72),
 ('Saint Pierre and Miquelon', 9.72, 7.42)]